adaptdl/adaptdl/torch/adascale.py

```diff
@@ -249,15 +298,25 @@ def step(self, *args, **kwargs):
             args: Positional arguments passed to ``optimizer.step``.
             kwargs: Keyword arguments passed to ``optimizer.step``.
         """
+        if (self.is_accumulation_step()):
+            self._current_grad_acc_step += 1
+            self._made_step = False
+            return
+
+        self._current_grad_acc_step = 0
+        self._made_step = True
         initial_lr = [pg["lr"] for pg in self._optimizer.param_groups]
         offset = 0
         for param_group in self._optimizer.param_groups:
             size = len(param_group["params"])
             grad_sqr = self._state["norm_avg"][offset:offset + size].sum()
             grad_var = self._state["var_avg"][offset:offset + size].sum()
-            gain = (grad_var + grad_sqr) / (grad_var / self._scale + grad_sqr)
+            gain = (grad_var + grad_sqr) / (grad_var / self.scale + grad_sqr)
             param_group["lr"] = gain * param_group["lr"]
             offset += size
+        for group in self._optimizer.param_groups:
+            for p in group["params"]:
+                p.grad /= self._grad_acc_steps
```

Comment:
Any ideas how the user can apply gradient clipping with auto-accumulation? If they clip before `optimizer.step()` then they will be clipping the pre-averaged gradients. Maybe it's better to divide the gradient in the backward hook instead?

Reply:
Fixed: moved to backward hook